### ISSUE TITLE
RRuleSet and RRuleProperties both have `dt_start`

### ISSUE THREAD

**Issue by omid** [fmeringdal/rust-rrule#45]
IMHO, RRule implementation should be as much as possible similar to the iCalendar standard.
And based on this, RRule struct should not have dt_start property and it should be there in RRuleSet.

Now my question is that why do we have it in two places? Is it something we should fix in the API?

**Comment by ralpha** [comment]
An RRule needs a starting date in order to start the iterator. Otherwise it does not know where to start. Also the start date is related to the way the RRule internally works.
There are some properties that are also set on finalization to make everything work correctly. So the RRule, once finalized, is **not** the same as `RRULE:...` String itself but in struct form.
https://github.com/fmeringdal/rust-rrule/blob/bf657fbd16f65ab3eb4e1139a64cf4467f8c0663/rrule/src/parser/rrulestr.rs#L145

We could create an RRule without a start date, but the start date has to be know once the iterator is created and the RRule does need to finalize at that moment.
I just kept this part the same (as it was) during my last refactoring as this would entail even bigger changes.

So short, yes, it can be done. But we then have to move the date and finalization to right before the creation of the iterator.

Do you want to have it the same as standard because of serialization/deserialization? Or because of something else?

**Comment by omid** [comment]
Exactly the point is ser/de. (We still don't have it, but we cannot have it)
Actually, because serde is not there yet, I wanted to validate an "rrule" syntax by parse_str it, but I needed to add a random dt_start to it. But anyway, even if we had serde, I couldn't use it, since it needed dt_start. 

Currently this library is not a one to one iCalendar equivalent.

Now RRuleSet is kind of equivalent to the iCalendar standard. (just not that mature)
And RRule/RRuleProperties is not equivalent of iCalendar RRule.

I think:

1. We don't need both RRule and RRuleProperties. (or I don't get the reason of having "RRuleProperties" and "RRule".)
2. Why RRule needs to be iterative? and not _just_ RRuleSet? Even if we need to have iterator on RRule, I think we can provide the start time to get a new iterator, like: `(RRule::new())->get_iter(dt_start)`

wdyt?

**Comment by ralpha** [comment]
I see what you mean.

The reason behind `RRule` and `RRuleProperties` is that `RRuleProperties` is an unvalidated rrule string.
When you convert it to `RRule` is will validate it. At that point we know the `RRule` is valid and can be executed.
The validation step will return an error if it was not valid.

If an object should follow the spec it should be `RRuleProperties`. This is because it allows for serialization and deserialization.
I think it is possible to remove `dt_start` from there and just add it as a separate argument when doing `RRule::new(...)`.
Then we can implement serialize and deserialize for `RRuleProperties`.

Changing the `RRule` itself and removing the `dt_start` is not possible. Think of `RRule` more as an internal object for converting the spec into the creation of dates (using an iterator).

> We don't need both RRule and RRuleProperties. (or I don't get the reason of having "RRuleProperties" and "RRule".)

I hope the explanation above helps. If not let me know.

> Why RRule needs to be iterative?

This is just the way the date parser was implemented. It keeps the state of the last iterator to generate the next date.

**For example:** Lets say you want to have an RRule that returns every 4 weeks. The simplest way to do this is to get the start date and add 4 weeks to it. Then for the one after that you just add 4 weeks to the previous date. etc.
If you would want to instead calculate this from the start date every time again, you will enter a world of pain because of everything, like timezones, daylight savings, leap years, ...
This is also the reason the [`rrule.between(...)` function](https://github.com/fmeringdal/rust-rrule/blob/main/rrule/src/core/rrule.rs#L81:L86) basically start the iterator loops over all the days until it finds the start date and then start collecting the dates.

> Even if we need to have iterator on RRule, I think we can provide the start time to get a new iterator, like: (RRule::new())->get_iter(dt_start)

Like stated above, better to change `RRuleProperties` instead.
Also the [`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) trait does not allows for the passing of more arguments.

## So in short:
We should just change `RRuleProperties` and change the `RRule::new(...)` function to take in the start date.
`RRuleProperties` can then implement serialize and deserialize.

If you have any questions, please ask!

**Comment by omid** [comment]
Thanks @ralpha, a lot more clear now.
And I completely agree to move dt_start out of RRule.

I still have some questions :)
- Why do we need to have two versions of validated RRule and not-yet-validated version? Just performance?
- Why cannot we merge RRule and RRuleSet? Something like
```pub struct RRuleSet {
    pub rrule: Vec<RRuleProperties>,
    pub rdate: Vec<DateTime>,
    pub exrule: Vec<RRuleProperties>,
    pub exdate: Vec<DateTime>,
    pub dt_start: DateTime,
}
```

**Comment by ralpha** [comment]
> And I completely agree to move dt_start out of RRule.

I hope you mean `RRuleProperties`, not `RRule`.

> Why do we need to have two versions of validated RRule and not-yet-validated version? Just performance?

This is so you can create `RRuleProperties` as follows:
```rust
RRuleProperties {
    freq: Frequency::Weekly,
    count: Some(20),
    dt_start: ymd_hms(1997, 9, 2, 9, 0, 0),
    week_start: Weekday::Sun,
    ..Default::default()
}
```
This is a valid rrule (when looking at spec) but this would not actually work with the way the iterator works.
This is the reason I said look at `RRule` as just an library specific struct, not a spec thing.

When validator checks `RRuleProperties` it will change it and add the following property: `properties.by_weekday`
So the `RRuleProperties` given to `RRule::new(..)` will be different from the one stored in `RRule` itself.
https://github.com/fmeringdal/rust-rrule/blob/main/rrule/src/parser/rrulestr.rs#L136:L137

The `RRule` itself is a restricted item, meaning that once you called `RRule::new(..)` and you have and the struct. You can not change the `properties` contained in it. (aka read-only)
Note the lack of `pub` in front of `properties` [here](https://github.com/fmeringdal/rust-rrule/blob/main/rrule/src/core/rrule.rs#L9)
Note that `RRule` check much more: see https://github.com/fmeringdal/rust-rrule#validation-limits and https://github.com/fmeringdal/rust-rrule#safety-and-security

So `RRuleProperties` you can change by itself any time you like and make it not follow the spec or add just nonsense.
But once you feed it into `RRule::new(..)` it is validated and only accepted if it was valid or can be made valid.

This way you always know that if you have an `RRule` it will work and not just do something incorrect.

So this is way to protect the user, algorithm and system from itself. Otherwise weird things can happen.
There is really no other way to do this without compromises or limitation on flexibility.

- `RRule`: We are sure it follows spec and can be used to generate dates. (but is read-only)
- `RRuleProperties`: Just the structure of the spec, you can change any time you want in any way. But can not be guarantied to follow spec.

>  Just performance?

No, has nothing to do with performance. Just a way to make sure things work as they should and allow flexibility at the same time.

> Why cannot we merge RRule and RRuleSet? Something like ...

Because the `RRuleSet` is different from `RRule` in thing it accepts. (`RRuleSet` adds more features)
With the `RRuleSet` you created above, it could start the iterator, have a bogus `RRuleProperties` (because was not validated)
And might work for the first 2 iterations. And then return dates that is should have never returned. And thus be incorrect.

So just to make things short and clear:
Merge `RRule` and `RRuleProperties`: Not going to happen. Will break everything in unpredictable ways or we would have to limit `RRuleProperties` and thus remove the ability to create it using `RRuleProperties { ... }` and only allow it to be created using a function that has all the properties as arguments in 1 go.

But we can `RRuleProperties` and remove `dt_start` and add this to the function `RRule::new(...)`. `RRuleProperties` can then implement serialize and deserialize.

I don't see an immediate other (easy) way to solve this and make all users happy.

**Comment by omid** [comment]
> I hope you mean RRuleProperties, not RRule.

Oh, yes, I meant that :D sorry.
I'm trying to apply my changes and see you opinions.

But I still not convinced with this:

> Why cannot we merge RRule and RRuleSet? Something like ...

The future RRule will be like this:
```
RRule {
    properties: RRuleProperties,
    tz: Tz,
    dt_start: DateTime,
}
```

Isn't it the exact same thing with the following simple _future_ RRuleSet:
```
RRuleSet {
    rrule: vec!<RRuleProperties>,
    rdate: vec![],
    exrule: vec![],
    exdate: vec![],
    dt_start: DateTime,
    tz: Tz,
}
```

Why aren't they interchangable?

**Comment by omid** [comment]
Could you please check #46 ?
It's just related to put dt_start and tz to RRule.
I also changed the "RRule::new" to "RRuleProperties::build".

I don't like the naming of RRuleProperties, becuase it's both a builder and a storage!
I tried to rename it to RRuleBuilder, but then, implementing serde for the Builder will be strange :)

And my question in the last comment is still valid and not related to the PR #46 .

**Comment by omid** [comment]
A kind reminder @ralpha @fmeringdal :)

**Comment by ralpha** [comment]
> A kind reminder @ralpha @fmeringdal :)

Sorry, just been busy with a lot of things. I'll keep this tab open so I don't forget. Hopefully I'll have some time soon to take look at this.

**Comment by omid** [comment]
Just merged master into the branch. and fixed conflicts.

**Comment by ralpha** [comment]
Hey @omid this is still on my mind. But just don't have the time for it at the moment.
And it looks like in the next month or so I'll have even less time. So sorry to keep you (and everyone else) waiting.
I hope to still do this task at some point. But it might take a long while until I have more free time I can spend on this (there are also a lot of other open source projects that demand my attention).

**Comment by omid** [comment]
Thanks @ralpha. So maybe @fmeringdal can help checking this.

**Comment by fmeringdal** [comment]
Hey @omid! Really sorry for being so late to this conversation, it has been some busy times for me lately so I haven't been able to allocate time for non-critical issues. I really appreciate your thoughts and work on making this library easier to use.
Here are my thoughts of the different issues brought up here:

**RRule and RRuleProperties seperateion**
The library has become a lot more stable since the introduction of the `RRuleProperties` type which really makes the seperation of concerns much clearer in the codebase. @ralpha made some great points above regarding this type, and it will continue to stay the same.

**dt_start moving from RRuleProperties to RRule**
Yes, let us do that. Then `RRuleProperties` will be according to the spec and we can implement `Serialize` and `Deserialize` for it. Thanks a lot for you PR @omid :pray: It has some merge conflicts which needs to be resolved, which is my bad since I have not been to review it for a long time. I will give it a review soon (within next week hopefully), and we can get it merged.

**Should we merge RRule and RRuleSet?**
I think we should keep both types. These are some of the reasons why I think so:
- This library is largely a port of the javascript and python rrule implementation, they separate these types.
- I imagine that if we removed `RRule`, then we will need to complicate the API even more if the user wants to disable `EXDATES`,  `EXRULEs` and `RDATES`. I also think that some users not familiar with the spec might be surprised by the fact that the string parsing will allow these "exotic variants" and lead to unexpected behaviors if they accept strings from end users.

**Comment by omid** [comment]
Hey @fmeringdal, nice, I entirely forgot what I've done since December and now a big conflict :|
Anyway, I think I've fixed the conflicts, and it's fine now again.
